Source/DafnyLanguageServer/Handlers/DafnySemanticTokensHandler.cs

```diff
@@ -0,0 +1,77 @@
+﻿using Microsoft.Dafny.LanguageServer.Workspace;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.Language.SemanticTokens;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Boogie;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+
+namespace Microsoft.Dafny.LanguageServer.Handlers {
+  public class DafnySemanticTokensHandler : SemanticTokensHandlerBase {
+    private readonly ILogger logger;
+    private readonly IDocumentDatabase documents;
+
+    public DafnySemanticTokensHandler(ILogger<DafnySemanticTokensHandler> logger, IDocumentDatabase documents) {
+      this.logger = logger;
+      this.documents = documents;
+    }
+
+    protected override SemanticTokensRegistrationOptions CreateRegistrationOptions(SemanticTokensCapability capability,
+      ClientCapabilities clientCapabilities) {
+      return new SemanticTokensRegistrationOptions {
+        DocumentSelector = DocumentSelector.ForLanguage("dafny"),
+        Legend = new SemanticTokensLegend() { // FIXME adjust depending on client (capability.TokenModifiers / capability.TokenTypes)
+          TokenModifiers = new Container<SemanticTokenModifier>(SemanticTokenModifier.Defaults),
+          TokenTypes = new Container<SemanticTokenType>(SemanticTokenType.Defaults),
+        },
+        Full = new SemanticTokensCapabilityRequestFull {
+          Delta = true
+        },
+        Range = true
+      };
+    }
+
+    private void CollectBasicTokens(DafnySemanticTokensBuilder builder, Dafny.Program program) {
```

Comment:
`Basic` doesn't mean much to me, and I think `Semantic` in `CollectSemanticTokens` is incorrect because I would say parsing doesn't establish semantics, even though LSP mentions the word Semantic in many of its token related classes (that we are also using for the 'Basic' tokens).

I'm not sure about the names though. Maybe `PushTokensBasedOnLexing` and `PushTokensBasedOnParsing`

Reply:
Renamed to "CollectScannerTokens" and "CollectResolutionBasedTokens"